mypy/stubgen.py

```diff
@@ -1038,36 +1075,82 @@ def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
         if all(foundl):
             self._state = VAR
 
-    def is_namedtuple(self, expr: Expression) -> bool:
-        if not isinstance(expr, CallExpr):
-            return False
+    def is_namedtuple(self, expr: CallExpr) -> bool:
         callee = expr.callee
-        return (isinstance(callee, NameExpr) and callee.name.endswith("namedtuple")) or (
-            isinstance(callee, MemberExpr) and callee.name == "namedtuple"
+        return (
+            isinstance(callee, NameExpr)
+            and (self.refers_to_fullname(callee.name, "collections.namedtuple"))
+        ) or (
+            isinstance(callee, MemberExpr)
+            and isinstance(callee.expr, NameExpr)
+            and f"{callee.expr.name}.{callee.name}" == "collections.namedtuple"
         )
 
+    def is_typed_namedtuple(self, expr: CallExpr) -> bool:
+        callee = expr.callee
+        return (
+            isinstance(callee, NameExpr)
+            and self.refers_to_fullname(callee.name, TYPED_NAMEDTUPLE_NAMES)
+        ) or (
+            isinstance(callee, MemberExpr)
+            and isinstance(callee.expr, NameExpr)
+            and f"{callee.expr.name}.{callee.name}" in TYPED_NAMEDTUPLE_NAMES
+        )
+
+    def _get_namedtuple_fields(self, call: CallExpr) -> list[tuple[str, str]] | None:
+        if self.is_namedtuple(call):
+            items: list[str]
+            if isinstance(call.args[1], StrExpr):
+                items = call.args[1].value.replace(",", " ").split()
+            elif isinstance(call.args[1], (ListExpr, TupleExpr)):
+                items = []
+                for item in call.args[1].items:
+                    if isinstance(item, StrExpr):
+                        items.append(item.value)
+                    else:
+                        return None  # Invalid namedtuple field type
+            else:
+                return None  # Invalid namedtuple fields type
+            if items:
+                self.import_tracker.require_name("Incomplete")
+            return [(item, "Incomplete") for item in items]
+        elif self.is_typed_namedtuple(call):
+
+            if isinstance(call.args[1], (ListExpr, TupleExpr)):
+                fields: list[tuple[str, str]] = []
+                b = AliasPrinter(self)
+                for item in call.args[1].items:
+                    if isinstance(item, TupleExpr) and len(item.items) == 2:
+                        field_name, field_type = item.items
+                        if not isinstance(field_name, StrExpr):
+                            return None  # Invalid NamedTuple field name
+                        fields.append((field_name.value, field_type.accept(b)))
+                    else:
+                        return None  # Invalid NamedTuple field type
+                return fields
+            else:
+                return None  # Invalid NamedTuple fields type
```

Comment:
With early returns:
```suggestion
            fields_arg = call.args[1]
            if not isinstance(fields_arg, (ListExpr, TupleExpr)):
                return None

            fields: list[tuple[str, str]] = []
            b = AliasPrinter(self)
            for item in fields_arg.items:
                if not isinstance(item, TupleExpr) and len(item.items) == 2:
                    return None  # Invalid NamedTuple field type
                field_name, field_type = item.items
                if not isinstance(field_name, StrExpr):
                    return None  # Invalid NamedTuple field name
                fields.append((field_name.value, field_type.accept(b)))
            return fields
```

Reply:
BTW, on failure to parse, do we want to emit "Incomplete" or something rather than just ignore?

Reply:
Done. I also made the same change in the section before

Reply:
> BTW, on failure to parse, do we want to emit "Incomplete" or something rather than just ignore?

We emit Incomplete. It is done in the `process_namedtuple` method (this is the original behavior since before this PR).